File: go/serializer.go
package scl

import (
    "fmt"
    "strings"
)

type Serializer struct {
    indent int
}

func NewSerializer(indent int) *Serializer {
    return &Serializer{indent: indent}
}

func (s *Serializer) Serialize(data map[string]interface{}) (string, error) {
    return s.serialize(data, 0), nil
}

func (s *Serializer) serialize(data map[string]interface{}, level int) string {
    lines := make([]string, 0)
    indentStr := strings.Repeat(" ", s.indent*level)

    for key, value := range data {
        line := fmt.Sprintf("%s%s :: ", indentStr, key)

        switch v := value.(type) {
        case bool:
            if v {
                line += "bool { true }"
            } else {
                line += "bool { false }"
            }
        case int:
            line += fmt.Sprintf("num { %d }", v)
        case float64:
            line += fmt.Sprintf("fl { %v }", v)
        case string:
            if strings.Contains(v, "\n") {
                line += fmt.Sprintf("ml {\n%s    '%s'\n%s}", indentStr, v, indentStr)
            } else {
                escaped := strings.ReplaceAll(v, "\\", "\\\\")
                escaped = strings.ReplaceAll(escaped, "\"", "\\\"")
                line += fmt.Sprintf("str { \"%s\" }", escaped)
            }
        case map[string]interface{}:
            line += "class {\n"
            line += s.serialize(v, level+1)
            line += fmt.Sprintf("%s}", indentStr)
        case []interface{}:
            if len(v) == 0 {
                line += "list(str) { }"
            } else {
                first := v[0]
                var typeName string
                var items []string

                switch first.(type) {
                case bool:
                    typeName = "bool"
                    for _, item := range v {
                        if b, ok := item.(bool); ok {
                            if b {
                                items = append(items, "true")
                            } else {
                                items = append(items, "false")
                            }
                        } else {
                            return fmt.Sprintf("Mixed types in list for key '%s': expected all bool", key)
                        }
                    }
                case int:
                    typeName = "num"
                    for _, item := range v {
                        if i, ok := item.(int); ok {
                            items = append(items, fmt.Sprintf("%d", i))
                        } else {
                            return fmt.Sprintf("Mixed types in list for key '%s': expected all int", key)
                        }
                    }
                case float64:
                    typeName = "fl"
                    for _, item := range v {
                        switch n := item.(type) {
                        case float64:
                            items = append(items, fmt.Sprintf("%v", n))
                        case int:
                            items = append(items, fmt.Sprintf("%d", n))
                        default:
                            return fmt.Sprintf("Mixed types in list for key '%s': expected all numeric", key)
                        }
                    }
                case string:
                    typeName = "str"
                    for _, item := range v {
                        if str, ok := item.(string); ok {
                            escaped := strings.ReplaceAll(str, "\\", "\\\\")
                            escaped = strings.ReplaceAll(escaped, "\"", "\\\"")
                            items = append(items, fmt.Sprintf("\"%s\"", escaped))
                        } else {
                            return fmt.Sprintf("Mixed types in list for key '%s': expected all str", key)
                        }
                    }
                default:
                    return fmt.Sprintf("Unsupported list element type: %T", first)
                }
                line += fmt.Sprintf("list(%s) { %s }", typeName, strings.Join(items, ", "))
            }
        default:
            line += s.serializeDynamic(v)
        }
        lines = append(lines, line)
    }

    result := strings.Join(lines, "\n")
    if level == 0 {
        result += "\n"
    } else {
        result += "\n"
    }
    return result
}

func (s *Serializer) serializeDynamic(value interface{}) string {
    switch v := value.(type) {
    case bool:
        if v {
            return "dynamic { true }"
        }
        return "dynamic { false }"
    case int:
        return fmt.Sprintf("dynamic { %d }", v)
    case float64:
        return fmt.Sprintf("dynamic { %v }", v)
    case string:
        if strings.Contains(v, "\n") {
            return fmt.Sprintf("ml {\n    '%s'\n}", v)
        }
        escaped := strings.ReplaceAll(v, "\\", "\\\\")
        escaped = strings.ReplaceAll(escaped, "\"", "\\\"")
        return fmt.Sprintf("dynamic { \"%s\" }", escaped)
    default:
        return fmt.Sprintf("Unsupported value type: %T", v)
    }
}

File: rust/src/serializer.rs
use std::collections::HashMap;

use crate::errors::ParseError;
use crate::parser::Value;

pub struct Serializer {
    indent: usize,
}

impl Serializer {
    pub fn new(indent: usize) -> Self {
        Self { indent: if indent == 0 { 4 } else { indent } }
    }

    pub fn serialize_map(&self, data: &HashMap<String, Value>) -> String {
        let mut out = String::new();
        out.push_str(&self.serialize_level(data, 0));
        out.push('\n');
        out
    }

    fn serialize_level(&self, data: &HashMap<String, Value>, level: usize) -> String {
        let mut lines: Vec<String> = Vec::new();
        let indent = " ".repeat(self.indent * level);
        for (k, v) in data {
            let mut line = String::new();
            line.push_str(&indent);
            line.push_str(k);
            line.push_str(" :: ");
            line.push_str(&self.serialize_value(v, level));
            lines.push(line);
        }
        lines.join("\n")
    }

    fn serialize_value(&self, v: &Value, level: usize) -> String {
        match v {
            Value::Bool(b) => format!("bool {{ {} }}", if *b { "true" } else { "false" }),
            Value::Num(n) => format!("num {{ {} }}", n),
            Value::Fl(f) => format!("fl {{ {} }}", trim_float(*f)),
            Value::Str(s) => {
                if s.contains('\n') {
                    let indent = " ".repeat(self.indent * level);
                    format!("ml {{\n{}    '{}'\n{}}}", indent, s, indent)
                } else {
                    let esc = s.replace('\\', "\\\\").replace('"', "\\\"");
                    format!("str {{ \"{}\" }}", esc)
                }
            }
            Value::Ml(s) => {
                let indent = " ".repeat(self.indent * level);
                format!("ml {{\n{}    '{}'\n{}}}", indent, s, indent)
            }
            Value::Class(map) => {
                let mut s = String::from("class {\n");
                s.push_str(&self.serialize_level(map, level + 1));
                s.push('\n');
                s.push_str(&" ".repeat(self.indent * level));
                s.push('}');
                s
            }
            Value::ListBool(vs) => {
                let items = vs.iter().map(|b| if *b { "true".to_string() } else { "false".to_string() }).collect::<Vec<_>>().join(", ");
                format!("list(bool) {{ {} }}", items)
            }
            Value::ListNum(vs) => {
                let items = vs.iter().map(|n| n.to_string()).collect::<Vec<_>>().join(", ");
                format!("list(num) {{ {} }}", items)
            }
            Value::ListFl(vs) => {
                let items = vs.iter().map(|f| trim_float(*f)).collect::<Vec<_>>().join(", ");
                format!("list(fl) {{ {} }}", items)
            }
            Value::ListStr(vs) => {
                let items = vs.iter().map(|s| format!("\"{}\"", s.replace('\\', "\\\\").replace('"', "\\\""))).collect::<Vec<_>>().join(", ");
                format!("list(str) {{ {} }}", items)
            }
            Value::Dynamic(inner) => {
                self.serialize_dynamic(inner.as_ref(), level).expect("dynamic serialization failed")
            }
        }
    }

    pub fn serialize_dynamic(&self, v: &Value, level: usize) -> Result<String, ParseError> {
        let s = match v {
            Value::Bool(b) => format!("dynamic {{ {} }}", if *b { "true" } else { "false" }),
            Value::Num(n) => format!("dynamic {{ {} }}", n),
            Value::Fl(f) => format!("dynamic {{ {} }}", trim_float(*f)),
            Value::Str(s) => {
                if s.contains('\n') {
                    let indent = " ".repeat(self.indent * level);
                    format!("ml {{\n{}    '{}'\n{}}}", indent, s, indent)
                } else {
                    let esc = s.replace('\\', "\\\\").replace('"', "\\\"");
                    format!("dynamic {{ \"{}\" }}", esc)
                }
            }
            Value::Ml(s) => {
                let indent = " ".repeat(self.indent * level);
                format!("ml {{\n{}    '{}'\n{}}}", indent, s, indent)
            }
            Value::Class(_) | Value::ListBool(_) | Value::ListNum(_) | Value::ListFl(_) | Value::ListStr(_) => {
                return Err(ParseError::new("Unsupported value type for dynamic"));
            }
        };
        Ok(s)
    }
}

fn trim_float(f: f64) -> String {
    let s = format!("{}", f);
    if s.contains('.') {
        let s2 = s.trim_end_matches('0').trim_end_matches('.');
        if s2.is_empty() { "0".to_string() } else { s2.to_string() }
    } else {
        s
    }
}


